src/main/java/org/testng/IDataProviderMethod.java

```diff
@@ -0,0 +1,41 @@
+package org.testng;
+
+import java.lang.reflect.Method;
+import java.util.List;
+
+/**
+ * Represents the attributes of a {@link org.testng.annotations.DataProvider} annotated method.
+ */
+public interface IDataProviderMethod {
+    /**
+     * @return - The instance to which the data provider belongs to. <code>null</code> if the data provider
+     * is a static one.
+     */
+    Object getInstance();
+
+    /**
+     * @return - A {@link Method} object that represents the actual
+     * {@literal @}{@link org.testng.annotations.DataProvider} method.
+     */
+    Method getMethod();
+
+    /**
+     * The name of this DataProvider.
+     */
+    String getName();
+
+    /**
+     * Whether this data provider should be run in parallel.
+     */
+    boolean isParallel();
+
+    /**
+     * Which indices to run from this data provider, default: all.
+     */
+    List<Integer> getIndices();
+
+    /**
+     * @return the current {@link ITestContext}.
+     */
+    ITestContext getTestContext();
```

Comment:
I don't catch why we need the context here and I don't find where you use it. 
IMO, It should be provided outside the class because it has no relation with a data provider method and is global to all the run, not a specific instance.

Reply:
Yes, but current listeners provide it via the test result or the context itself.
I'm not feeling good to host it in the model: IDataProviderMethod is only supposed to define a data provider method (static) where the context depends on the execution (dynamic)

Reply:
Since we don't have an `ITestResult` for the data provider, that part is ruled out.
So the only option available here is to inject the `ITestContext` explicitly as a parameter for the listener. Something like this 

```java
void beforeDataProviderExecution(IDataProviderMethod dataProviderMethod, ITestNGMethod method, ITestContext iTestContext);
void afterDataProviderExecution(IDataProviderMethod dataProviderMethod, ITestNGMethod method, ITestContext iTestContext);
```